Introduce subdivisions aliases

sphh opened the pull request:
<!--
  Thanks for contributing to python-holidays!
-->

## Proposed change

<!--
  Describe the big picture of your changes.
  Don't forget to link your PR to an existing issue if any.
-->

Adds aliases for subdivision regions. Closes #1660

## Type of change

<!--
  Type of change you want to introduce. Please, check one (1) box only!
  If your PR requires multiple boxes to be checked, most likely it needs to
  be split into multiple PRs.
-->

- [ ] New country/market holidays support (thank you!)
- [ ] Supported country/market holidays update (calendar discrepancy fix, localization)
- [ ] Existing code/documentation/test/process quality improvement (best practice, cleanup, refactoring, optimization)
- [ ] Dependency update (version deprecation/upgrade)
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Breaking change (a code change causing existing functionality to break)
- [x] New feature (new python-holidays functionality in general)

## Checklist

<!--
  Put an `x` in the boxes that apply. You can change them after PR is created.
-->

- [ ] I've followed the [contributing guidelines][contributing-guidelines]
- [ ] I've run `make pre-commit`, it didn't generate any changes
- [ ] I've run `make test`, all tests passed locally

Updating of the documentation and adding test cases still needs to be done. I just wanted to check, if you are happy with my implementation.

<!--
  Thanks again for your contribution!
-->

[contributing-guidelines]: https://github.com/vacanza/python-holidays/blob/beta/CONTRIBUTING.rst
[docs]: https://github.com/vacanza/python-holidays/tree/beta/docs/source

arkid15r commented:
Thanks for the PR @sphh!

As I mentioned earlier tests are very important in order to keep the library stable. 
First you need to make sure your code doesn't break anything (see the PR checklist). 
Then considering significance of the change I'd expect the new functionality to be covered with tests (see the hints I gave at the corresponding issue). Let me know if you need help with that.

sphh commented:
Yes, I am with you. But as we have already started at the back (API has to be defined and tests should be written first), could you please tell me, if the API in added is something you can live with:
1. A class level attribute `subdivisions_aliases` with the alias as key and the official code as value?
2. Storing the subdivision parameter in `self.subdiv` as it has been given by the user and if it is an alias only resolving it at the very last moment?

If that is ok with you and your ideas, I can do the tedious tasks of adding tests (if you don't want to do it) and updating the documentation (again if you as assumed native speaker are not better suited to do it).

arkid15r commented:
> Yes, I am with you. But as we have already started at the back (API has to be defined and tests should be written first), could you please tell me, if the API in added is something you can live with:
> 
> 1. A class level attribute `subdivisions_aliases` with the alias as key and the official code as value?
> 2. Storing the subdivision parameter in `self.subdiv` as it has been given by the user and if it is an alias only resolving it at the very last moment?
> 
> If that is ok with you and your ideas, I can do the tedious tasks of adding tests (if you don't want to do it) and updating the documentation (again if you as assumed native speaker are not better suited to do it).

Yes, those 2 points sound good to me. Looking forward to reviewing the complete version of the code!

Thank you for implementing this functionality!

sphh commented:
That's great. Thanks for confirming!

I have another question, regarding the table in the section [Available countries](https://github.com/vacanza/python-holidays#user-content-available-countries), which is down to personal liking: Where do you want to have the additional column with the subdivision aliases? Between the ‘Subdivisions’ and ‘Supported Languages’ columns or as last column. Personally I can see it in both places …

arkid15r commented:
> That's great. Thanks for confirming!
> 
> I have another question, regarding the table in the section [Available countries](https://github.com/vacanza/python-holidays#user-content-available-countries), which is down to personal liking: Where do you want to have the additional column with the subdivision aliases? Between the ‘Subdivisions’ and ‘Supported Languages’ columns or as last column. Personally I can see it in both places …

This one needs more attention. I don't think we need to add a separate column for this. I expect the subdivisions aliases to be a very useful from UX perspective but not widely used feature. Maybe it'd be a good idea to keep it simple and add subdiv alias right to its ISO3166-2 value in brackets, e.g. `1 (B), 2 (K)...`. I would expect 1 alias per subdiv (maybe 2 at top in really weird cases).

sphh commented:
Sounds like a sensible plan!

sphh commented:
The last commits update the README, adds or updates tests and adds one utility function.

The tests flag up two problems:
1. `make pre-commit` complains about missing documents. Honestly I have no clue what to do since I have not touched these documents.
2. `make test` complains about missing Ukrainian translation. Here I am out of my wits, because I do not speak or read Ukrainian.

I updated the test for subdivisions in table in the README, but there is no test to check, if the given aliases actually exist. Can I leave that to you? I frankly admit, that I am definitely not good at writing tests …

KJhellico reviewed (commented):
@sphh, this is an impressive contribution! Please look at some suggestions (Ukrainian translation mainly).

sphh commented on tests/countries/test_austria.py:
```diff
@@ -148,6 +149,7 @@ def test_bank_2022(self):
         self.assertHolidays(
             Austria(categories=BANK, years=2022),
             ("2022-04-15", "Karfreitag"),
+            ("2022-11-15", "Hl. Leopold"),
```
Sounds like a good idea. If @arkid15r is happy with it, I can have a look into it.

sphh commented on holidays/countries/austria.py:
```diff
@@ -86,6 +130,46 @@ def _populate_bank_holidays(self):
+
+    def _populate_subdiv_2_bank_holidays(self):
+        # Patron saint of Carinthia: Hl. Josef (St. Joseph's day)
+        self._add_holiday_mar_19(tr("Hl. Josef"))
```
Thanks for the translations!

Thanks for pointing out the popularity of St. Joseph; enough for him to get a dedicated day in `python-holiday`!

sphh commented:
@KJhellico: Can you or some other guru help with the failed checks? Looking at them I get the impression, that tests fail for files I have never touched.

arkid15r reviewed (commented):
I've spent some time on making the code complaint with our general standards:
  - normally we don't change file headers (see recent https://github.com/vacanza/python-holidays/pull/1672)
  - narrow down AU subdiv aliases
  - add tests to keep the coverage at 100% (found a typo bug while was working on that)
  - move `_get_subdiv_aliases` from utils to HolidayBase
  - update doc tests with subdivision aliases part

Let's take another look at it!

arkid15r commented:
> I strongly believe to give credit to whom it is due. With this type of projects that means to include a copyright notice. I think, that is only fair. I also know, that many software developers do not like to have these notices in the headers of their files. (Some even request that the contributors pass on their copyright to the project holder!). So I find this approach is a decision worthy of King Solomon! Thanks

That's a fair concern. I don't have opinion on how it should be implemented for now. I've been just following previous practices. That's why I pointed out normally the headers are not to be modified. If you have ideas on how to implement copyright/credits related stuff properly please create and issue for further discussion.

> @arkid15r: Can you please explain the reason for removing the other aliases, e.g. 'Bgl' and 'Burgenland' (here the abbreviation and the full name) for subdivision 'AT-1'? This follows the country approach of having the three letter and also the full name beside the two-letter identifier for the holidays, e.g. 'AUT' and 'Austria' beside 'AT'?

I don't mind keeping those aliases, just not all of them. Let's pick a minimum set to resolve the initial problem of AT irrelevant subdivision names. I think using full names as they exist in German language (I believe) would be the best. It's the most obvious way to me. Not really sure about 2-3 letter identifiers -- those are not ISO 3166 codes so we don't need to "create" them. For aliases we can pick anything and I want to make sure we pick just enough.

> @arkid15r: I don't mind at all moving that functionality to `HolidayBase` as long as it is not a private method.
> 
> With enforcing only one alias for each subdivision (which personally I do not like very much – see above) this method does not give any added functionality compared to use the class attribute `HolidayBase.subdivisions_aliases`. Actually you loose functionality, because in `subdivisions_aliases` it is possible to directly look up the ISO 3166-2 subdivision code, which is not possible with the tuple returned by `get_subdiv_aliases()`!
> 
> The functionality of `list_supported_aliases()` lies in the ability to directly look up (all) the aliases for an ISO 3166-2 code, which is now sadly lost.

That's a good point. I've refactored it back.

arkid15r commented on holidays/countries/austria.py:
```diff
@@ -21,16 +21,32 @@ class Austria(HolidayBase, ChristianHolidays, InternationalHolidays):
+        "9",  # Wien.
+    )
+    subdivisions_aliases = {
+        "B": "1",
+        "K": "2",
+        "N": "3",
+        "O": "4",
+        "S": "5",
+        "St": "6",
+        "T": "7",
+        "V": "8",
+        "W": "9",
```
```suggestion
        "Burgenland": "1",
        "Kärnten": "2",
        "Niederösterreich": "3",
        "Oberösterreich": "4",
        "Salzburg": "5",
        "Steiermark": "6",
        "Tirol": "7",
        "Vorarlberg": "8",
        "Wien": "9",
```

This is how imagine aliases resolving the initial AT irrelevant subdivision names problem.
@sphh what do you think would be a reasonable addition to this mapping?

arkid15r commented on holidays/countries/austria.py:
```diff
@@ -86,6 +102,46 @@ def _populate_bank_holidays(self):
+        self._add_holiday_nov_11(tr("Hl. Martin"))
+
+    def _populate_subdiv_2_bank_holidays(self):
+        # Patron saint of Carinthia: Hl. Josef (St. Joseph's day)
```
I suggest simplifying the comments to just their names in English for brevity.
```suggestion
        # St. Joseph's Day.
```

arkid15r commented:
> If I were asked, I am totally happy with your approach!

Yeah, we'll figure something out.

> You are right, none of the aliases are ISO 3166 codes. My personal approach would be to use all versions, which are in common (official?) use. In Austria's case these are also the abbreviations. But I think, at least a relevant subdivision name _and_ the full name should be included.

I've made a suggestion for the full names. They are 100% descriptive and understandable. If there is an official source for the used abbreviations/symbols of AT subdivisions I'd love to take a look. Adding shorter or/and more popular options would be beneficial.


> In the case of having just two aliases, shouldn't they be shown in the table of supported countries? Or at least indicate, that there are more aliases defined (that's what I did with the ellipsis)?

Yes, there should be no way to add an alias w/o updating the docs. I added a test for that but it doesn't seem to cover multiple alias cases.


> Great! Please also see my comments on the code itself.

What do you mean?

sphh commented:
> > If I were asked, I am totally happy with your approach!
> 
> Yeah, we'll figure something out.

I don't understand: Haven't you already figured it out?
 
> > You are right, none of the aliases are ISO 3166 codes. My personal approach would be to use all versions, which are in common (official?) use. In Austria's case these are also the abbreviations. But I think, at least a relevant subdivision name _and_ the full name should be included.
> 
> I've made a suggestion for the full names. They are 100% descriptive and understandable. If there is an official source for the used abbreviations/symbols of AT subdivisions I'd love to take a look. Adding shorter or/and more popular options would be beneficial.

This is the official document for the abbreviations I used for Austria: http://wiki.cacert.org/AustrianCommunity/abkuerzungen_laender_oesterreich

> > In the case of having just two aliases, shouldn't they be shown in the table of supported countries? Or at least indicate, that there are more aliases defined (that's what I did with the ellipsis)?
> 
> Yes, there should be no way to add an alias w/o updating the docs. 

I interpreted your https://github.com/vacanza/python-holidays/pull/1662#issuecomment-1908847471, that the README should only show one (short) alias for each subdivision. To give an indication, that there are more aliases available, I introduced the ellipsis ‘…’ (and that complicated regexp).

> I added a test for that but it doesn't seem to cover multiple alias cases.

If the approach with the ellipsis is adopted, IMHO the tests should on check, if the given alias exists and the ellipsis is justified.

> > Great! Please also see my comments on the code itself.
> 
> What do you mean?

Please see

- https://github.com/vacanza/python-holidays/pull/1662#review-thread-or-comment-id-939637231
- https://github.com/vacanza/python-holidays/pull/1662#review-thread-or-comment-id-939636121

arkid15r commented:
> I don't understand: Haven't you already figured it out?

No, not yet. We have WIP [PR](https://github.com/vacanza/python-holidays/pull/1672) by @dr-prodigy for yearly (c) update.

> This is the official document for the abbreviations I used for Austria: http://wiki.cacert.org/AustrianCommunity/abkuerzungen_laender_oesterreich

Doesn't seem to be "official grade" to me. But it's better than nothing. I think we can work with the subdivision short versions from there.

> I interpreted your [#1662 (comment)](https://github.com/vacanza/python-holidays/pull/1662#issuecomment-1908847471), that the README should only show one (short) alias for each subdivision. To give an indication, that there are more aliases available, I introduced the ellipsis ‘…’ (and that complicated regexp).

If we have 1-2 aliases I'd just put them all to the docs so there would be no need to look into the source code for all available aliases.

> Please see
> 
> * [Introduce subdivisions aliases #1662 (comment)](https://github.com/vacanza/python-holidays/pull/1662#review-thread-or-comment-id-939637231)
> * [Introduce subdivisions aliases #1662 (comment)](https://github.com/vacanza/python-holidays/pull/1662#review-thread-or-comment-id-939636121)

I saw that. Could you add a bit more details on what specifically I should look at? 
What exactly needs to be addressed?

Thank you!

sphh commented:
> No, not yet. We have WIP [PR](https://github.com/vacanza/python-holidays/pull/1672) by @dr-prodigy for yearly (c) update.

Ah, I see. I was of the impression (because of receiving the pull request in my branch), that this is already accepted!

> Doesn't seem to be "official grade" to me. But it's better than nothing. I think we can work with the subdivision short versions from there.

Well, this is referencing the ÖNORM, which are the Austrian's local standards (like DIN for Germany and ISO standards for international use). Can there be anything more ‘official’?

> If we have 1-2 aliases I'd just put them all to the docs so there would be no need to look into the source code for all available aliases.

But what to do, if we have more than 2 aliases?

> > Please see
> > 
> > * [Introduce subdivisions aliases #1662 (comment)](https://github.com/vacanza/python-holidays/pull/1662#review-thread-or-comment-id-939637231)
> > * [Introduce subdivisions aliases #1662 (comment)](https://github.com/vacanza/python-holidays/pull/1662#review-thread-or-comment-id-939636121)
> 
> I saw that. Could you add a bit more details on what specifically I should look at? What exactly needs to be addressed?

The first comment concerns line 686 in `holiday_base.py` (somehow I cannot manage to get a link to it). Shouldn't that line read
```python3
subdivision_aliases: Dict[str, List[str]] = {}
```
instead of
```python3
subdivision_aliases: Dict[str, List] = {}
```
(watch for the additional `str` in the annotation)?

The second comment concerns the whole method `HolidayBase.get_subdivision_aliases()` and the difference between your and my implementations. This difference is only noticeable, if there is one or more subdivisions which do not have an alias defined. The dictionary returned by your implementation is missing these subdivisions as keys. Whereas my implementation has _all_ available subdivisions as keys (with empty lists as values if no aliases are defined).

arkid15r commented:
> Well, this is referencing the ÖNORM, which are the Austrian's local standards (like DIN for Germany and ISO standards for international use). Can there be anything more ‘official’?

From my perspective an official source would be somehow related to AT government (.gv.at I guess).

>But what to do, if we have more than 2 aliases?

Include those too. As I said I don't expect more than 2 aliases per subdiv.

>HolidayBase.get_subdivision_aliases()

Let's use your implementation 👍

sphh commented:
> > Well, this is referencing the ÖNORM, which are the Austrian's local standards (like DIN for Germany and ISO standards for international use). Can there be anything more ‘official’?
> 
> From my perspective an official source would be somehow related to AT government (.gv.at I guess).

Mmh, not even the ISO would then rate as an official source: “The International Organization for Standardization (ISO […]) is an independent, **non-governmental**, international standard development organization” (my emphasis, https://en.wikipedia.org/wiki/International_Organization_for_Standardization).

There are many national standard organisations: [DIN](https://en.wikipedia.org/wiki/Deutsches_Institut_f%C3%BCr_Normung) for Germany, [bsi](https://en.wikipedia.org/wiki/BSI_Group) for UK, [ANSI](https://en.wikipedia.org/wiki/American_National_Standards_Institute) for the US and (ÖNORM](https://en.wikipedia.org/wiki/Austrian_Standards_International) just to name a few. And there is [CEN](https://en.wikipedia.org/wiki/European_Committee_for_Standardization) for the EU. See also https://en.wikipedia.org/wiki/Standards_organization.

> > But what to do, if we have more than 2 aliases?
> 
> Include those too. As I said I don't expect more than 2 aliases per subdiv.

For Austria we already have 3. But that might be a special case because of the unfortunate naming in the ISO 3166-2 standards …

> > HolidayBase.get_subdivision_aliases()
> 
> Let's use your implementation 👍

:+1:

arkid15r commented:
> Mmh, not even the ISO would then rate as an official source: “The International Organization for Standardization (ISO […]) is an independent, non-governmental, international standard development organization” (my emphasis, https://en.wikipedia.org/wiki/International_Organization_for_Standardization).
> 
> There are many national standard organisations: [DIN](https://en.wikipedia.org/wiki/Deutsches_Institut_f%C3%BCr_Normung) for Germany, [bsi](https://en.wikipedia.org/wiki/BSI_Group) for UK, [ANSI](https://en.wikipedia.org/wiki/American_National_Standards_Institute) for the US and (ÖNORM](https://en.wikipedia.org/wiki/Austrian_Standards_International) just to name a few. And there is [CEN](https://en.wikipedia.org/wiki/European_Committee_for_Standardization) for the EU. See also https://en.wikipedia.org/wiki/Standards_organization.

So based on this reasoning the iso.org is kind of equal to cacert.org? I'm not going to argue with you on this. However, checking with gov related sites has been our standard verification practice. The idea was to find those aliases used on official Austria sites. I've found [this example](https://www.bfa.gv.at/402/files/01_Broschueren/Informationsbroschuere_Asylverfahren_in_Oesterreich_EN.pdf) and it's different from the one above.

I don't like the idea of putting unverified abbreviations into the alias list. The long story short: if there is a proposal to add an alias that is different from the subdivision's full name It'd really beneficial to see the source or/and example of official usage.

This is my current understanding of existing mapping which should be added together with the full names (see the code suggestion):
   ```
    Burgenland - Bgld.
    Carinthia (Kärnten) - Ktn.
    Lower Austria (Niederösterreich) - NÖ
    Upper Austria (Oberösterreich) - OÖ
    Salzburg - Salzbg. or Sbg.
    Styria (Steiermark) - Stmk.
    Tyrol (Tirol) - T.
    Vorarlberg - Vbg. or Vor.
    Vienna (Wien) - W.
```


> For Austria we already have 3. But that might be a special case because of the unfortunate naming in the ISO 3166-2 standards

The current PR code contains just 1 alias per subdivision with a pending suggestion to change them to their full names.

sphh commented:
> So based on this reasoning the iso.org is kind of equal to cacert.org?

No, cacert.org is just quoting an ÖNORM, which is the national equivalent to the ISO. As is DIN, ANSI, BS or any other national standard.

> I've found [this example](https://www.bfa.gv.at/402/files/01_Broschueren/Informationsbroschuere_Asylverfahren_in_Oesterreich_EN.pdf) and it's different from the one above.

Yes, it is. But then they define it at the beginning and thus they can use whatever they like…

> I don't like the idea of putting unverified abbreviations into the alias list. The long story short: if there is a proposal to add an alias that is different from the subdivision's full name It'd really beneficial to see the source or/and example of official usage.

And here comes the ÖNORM in!

I would say, the aliases should contain the full name, a one (or two-letter) designator without a period at the end and an abbreviation, hence
```
Burgenland - B - Bgld.
Carinthia (Kärnten) - K - Ktn.
Lower Austria (Niederösterreich) - N - NÖ
Upper Austria (Oberösterreich) - O - OÖ
Salzburg - S - Sbg.
Styria (Steiermark) - S - Stmk.
Tyrol (Tirol) - T
Vorarlberg - V- Vbg.
Vienna (Wien) - W
```
The English full name can be provided by the translation.

arkid15r commented:
> And here comes the ÖNORM in!

You keep mentioning this as solution for the problem. Can I see a link to their standard regarding AT state names or whatever you use for your reasoning?

I'm having a [hard time to find examples of real use cases](https://www.statistik.at/atlas/blick/#) for the aliases we've been discussing for a while. Is there anything that can help answer, for example, this question: "why B / Bgld for Burgenland?"

>The English full name can be provided by the translation.

I don't think we want to do the aliases localization. We have ISO subdivision codes for everyone. 
Do you think it's something we should necessarily implement?

sphh commented:
> > And here comes the ÖNORM in!
> 
> You keep mentioning this as solution for the problem.

Yes, because the ÖNORM (or DIN or BS or ANSI) are the national versions of the ISO. I reference this, because in the beginning you insisted that everything has to be mentioned in standards. May I ask, why the national standards are not acceptable, but the international is?

> Can I see a link to their standard regarding AT state names or whatever you use for your reasoning?

If you want, you can buy their standards. Or you trust cacert.org to cite it correctly. Or see below.

BTW have you tried to search for ‘AT-0’, the official ISO denomination? Did you get any results?

I am still not sure, what you need, so that you can accept an alias?

IMHO aliases are there to make life easier and there are many common aliases, which are in common use but not written down in laws or standards …

> I'm having a [hard time to find examples of real use cases](https://www.statistik.at/atlas/blick/#) for the aliases we've been discussing for a while. Is there anything that can help answer, for example, this question: "why B / Bgld for Burgenland?"

Here is my collection of supporting documents:

- There is an official dictionary in Austria. This is the final authoritative source for all matters regarding the Austrian version of the German language. It is published on behalf of the government and endorsed by it. I made a picture of the title page and the entries for Burgenland, Kärnten, Niederösterreich, Oberösterreich, Salzburg, Steiermark and Vorarlberg. Please see for yourself: 
[OeWB.pdf](https://github.com/vacanza/python-holidays/files/14180062/OeWB.pdf)
- There is an official Austrian website offering all legislative documents run by the government (www.ris.bka.gv.at). Here are some links to documents using all these abbreviations:
    - Page 2, 2.1.c) https://www.ris.bka.gv.at/Dokumente/Bundesnormen/NOR40206699/NOR40206699.pdf
    - Page 1, Table https://www.ris.bka.gv.at/Dokumente/Bundesnormen/NOR40102820/NOR40102820.pdf

Would these documents alleviate your reservations?

> > The English full name can be provided by the translation.
> 
> I don't think we want to do the aliases localization. We have ISO subdivision codes for everyone. Do you think it's something we should necessarily implement?

I only mentioned this, because you mentioned the English and the local German versions in one go. Please do not combine these into one alias!

arkid15r reviewed (commented):
I've updated this PR based on the information provided by @sphh. Thank you, that helped a lot with the subdiv aliases decision.

Let's work a bit more on this and get the change merged! I think we're very close. I'm a bit concerned with `test_docs.py` tests complexity though -- it needs to be simplified later.

@sphh, @KJhellico please review this code once again when you have a chance.

arkid15r commented:
> But why have the single letter aliases (B, K, N, O, S, St and V) been removed? I thought, it was decided, that they are ok …

I hadn't found any references to those aliases so I removed them from the initial version.

> Any reason why you don't take my approach with the single regexp? That makes the Python code simpler admittedly at the cost of a more complex regexp. Personally I find that approach easier to follow (especially if there is a comment, explaining what that regexp does). But I guess it is down to personal preferences (comments explaining the code always helps).

No, no reason. I picked up from the most recent code version. Feel free to change it back.

Please also note I forgot to revert `HolidayBase::get_subdivision_aliases` to use your implementation.

sphh commented:
> I hadn't found any references to those aliases so I removed them from the initial version.

Would a copy of the official dictionary showing these abbreviations suffice? If yes, I'll create it.

> No, no reason. I picked up from the most recent code version. Feel free to change it back.

I guess, you will find your commit, which changed it, faster than I do. Frankly, I already spent much too much time on all this for such a simple PR …

> Please also note I forgot to revert `HolidayBase::get_subdivision_aliases` to use your implementation.

Very good, I noticed that (but has that been changed some time ago?)

arkid15r commented:
> Would a copy of the official dictionary showing these abbreviations suffice? If yes, I'll create it.

I created the list based on the materials you've provided as well as my own research results. Yes, the dictionary was one of the sources I used. My point is still the same -- if an alias is in a wide use in AT I'd like to have an evidence of that in any type of (semi-) official documents, gov sites, other sources. I have a broader question: why 2 aliases (full name + short name) isn't enough?

Another question is that the list already contains two one-letter aliases -- how to deal with those cases? I'd like to have consistency at this level too. Ideally It'd be great to see a final state of the alias list from your perspective first.

> I guess, you will find your commit, which changed it, faster than I do. Frankly, I already spent much too much time on all this for such a simple PR …

The current code looks okay to me. 
 
> Very good, I noticed that (but has that been changed some time ago?)

Not sure what you mean. I see that the current implementation differs from your original code.

sphh commented:
This is the supporting document for the single letter abbreviations: 
[single_letter_abbreviations.pdf](https://github.com/vacanza/python-holidays/files/14220557/single_letter_abbreviations.pdf)

arkid15r commented:
> This is the supporting document for the single letter abbreviations: [single_letter_abbreviations.pdf](https://github.com/vacanza/python-holidays/files/14220557/single_letter_abbreviations.pdf)

Thanks for compiling this! I ended up with the following with my own attempt to figure this out:

> The two-letter abbreviations I provided earlier (e.g., Bg, K, NÖ, OÖ, S, Stmk, V, W) are more commonly used and widely recognized in Austria than single-letter abbreviations. They are typically used in postal addresses, official documents, and other written communication for brevity and clarity.
> 
> While single-letter abbreviations (e.g., B, K, N, O, S, St, V, W) might also be used informally or in specific contexts, they are less common compared to the two-letter abbreviations and may not be as universally understood.
> 
> In summary, the two-letter abbreviations are more frequently used and standardized for representing Austrian federal states in various forms of communication.

I still think 2 aliases would be sufficient for majority (maybe all) of cases. However, one of our lower priority goals is better user experience. So I believe this may be actually beneficial (according to at least one user from the target audience -- @sphh).

I hope this is the final list of aliases everybody agrees on.

I also updated `README.rst` and `HolidayBase::get_subdivision_aliases` code.

PTAL!

sphh commented:
> I ended up with the following with my own attempt to figure this out:
> 
> > The two-letter abbreviations I provided earlier (e.g., Bg, K, NÖ, OÖ, S, Stmk, V, W) are more commonly used and widely recognized in Austria than single-letter abbreviations. They are typically used in postal addresses, official documents, and other written communication for brevity and clarity.
> > While single-letter abbreviations (e.g., B, K, N, O, S, St, V, W) might also be used informally or in specific contexts, they are less common compared to the two-letter abbreviations and may not be as universally understood.
> > In summary, the two-letter abbreviations are more frequently used and standardized for representing Austrian federal states in various forms of communication.

I am afraid, that is not my perception how they are used here in Austria. It might be confusing, that there are more than one abbreviations for one single state, but do not worry, all of these abbreviations are used in one way or the other in official and informal documents. Thanks for including them all! I really trust this is beneficial – at least for all Austrians.

> I hope this is the final list of aliases everybody agrees on.

I am more than happy with that list of aliases from the commit https://github.com/vacanza/python-holidays/pull/1662/commits/6e85ec66a2778f8ec500acf86cc46b48b2182248 (Why wouldn't I, because these are exactly the aliases from my first commit…) Thanks.

arkid15r commented:
> I am more than happy with that list of aliases from the commit [6e85ec6](https://github.com/vacanza/python-holidays/commit/6e85ec66a2778f8ec500acf86cc46b48b2182248) (Why wouldn't I, because these are exactly the aliases from my first commit…) Thanks.

Not exactly the same but very close to it. Thanks for working with me on this @sphh 👍

@KJhellico could you review it when you get a chance?

KJhellico reviewed (dismissed):
LGTM! Just one suggestion, for l10n comments consistency.

arkid15r commented on holidays/countries/austria.py:
```diff
@@ -86,6 +119,46 @@ def _populate_bank_holidays(self):
+    def _populate_subdiv_4_bank_holidays(self):
+        # St. Florian's Day.
+        if self._year >= 2004:
+            self._add_holiday_may_4(tr("Hl. Florian"))
```
Updated, nice catch!
